### PR TITLE
Add json and csv formats to upgrade output

### LINKED ISSUE #2415: The "theme update" command does not output JSON
I am using the command `theme update` to preview the themes that will be updated, but it does not output a JSON result:
```
wp-cli theme update --all --path=/var/www/vhosts/mywordpress --json --dry-run
```

Then I tried the `--format` option with `table` and `summary` values. Both of them displays the same output, so it looks like there are some problems with the output part.

### PR DESCRIPTION
**Pull request by gilbitron**
Resolves #2415

### CODE REVIEW

**Comment by gilbitron** [comment]
@danielbachhuber While putting together this PR I realised that the output of `--format=json` and `--format=csv` should probably not include the progress updates. For example:

```
Enabling Maintenance mode...
Downloading update from https://downloads.wordpress.org/theme/twentyfifteen.1.4.zip...
Using cached file '/Users/gilbitron/.wp-cli/cache/theme/twentyfifteen-1.4.zip'...
Unpacking the update...
Installing the latest version...
Removing the old version of the theme...
Theme updated successfully.
Disabling Maintenance mode...
Success: Updated 1/1 themes.
```

Any ideas how we would get around this?

**Comment by danielbachhuber** [comment]
> should probably not include the progress updates

Correct.

> Any ideas how we would get around this?

Most of the output is from `WP_CLI\UpgraderSkin`. Given we can't pass much context to the callbacks, our next best bet would be to assign a class-level variable to the object.

This worked well in my testing:

```
diff --git a/php/WP_CLI/CommandWithUpgrade.php b/php/WP_CLI/CommandWithUpgrade.php
index f7da951..6c27d65 100755
--- a/php/WP_CLI/CommandWithUpgrade.php
+++ b/php/WP_CLI/CommandWithUpgrade.php
@@ -249,6 +249,7 @@ abstract class CommandWithUpgrade extends \WP_CLI_Command {
                                $cache_manager->whitelist_package($item['update_package'], $this->item_type, $item['name'], $item['update_version']);
                        }
                        $upgrader = $this->get_upgrader( $assoc_args );
+                       $upgrader->cli_output_format = ! empty( $assoc_args['format'] ) ? $assoc_args['format'] : '';
                        $result = $upgrader->bulk_upgrade( wp_list_pluck( $items_to_update, 'update_id' ) );
                }

diff --git a/php/WP_CLI/UpgraderSkin.php b/php/WP_CLI/UpgraderSkin.php
index aa56b43..b2b3d0f 100644
--- a/php/WP_CLI/UpgraderSkin.php
+++ b/php/WP_CLI/UpgraderSkin.php
@@ -41,7 +41,9 @@ class UpgraderSkin extends \WP_Upgrader_Skin {

                $string = str_replace( '&#8230;', '...', strip_tags( $string ) );

-               \WP_CLI::line( $string );
+               if ( ! empty( $this->cli_output_format ) && ! in_array( $this->cli_output_format, array( 'csv', 'json' ) ) ) {
+                       \WP_CLI::line( $string );
+               }
        }
 }
```

**Comment by gilbitron** [comment]
@danielbachhuber Almost there. However when using `--format=json` and `--format=csv` I'm still getting this single line of output before the json/csv:

```
Using cached file '/Users/gilbitron/.wp-cli/cache/theme/twentyfifteen-1.4.zip'...
```

Any ideas how to mute that?

**Comment by gilbitron** [comment]
Also my tests are looking for incomplete data because of the fact they don't know what the "upgrading to" version will be. Any ideas how to get around that? Is there a way we can find the "upgrade to" version and do something like `And save STDOUT as {UPGRADE_VERSION}`?

**Comment by danielbachhuber** [comment]
> However when using --format=json and --format=csv I'm still getting this single line of output before the json/csv:

Ok. Let's go with a different approach. Early in the `update()` method, let's do:

```
if ( ! empty( $assoc_args['format'] ) && in_array( $assoc_args['format'], array( 'csv', 'json' ) ) ) {
	$logger = new \WP_CLI\Loggers\Quiet;
	WP_CLI::set_logger( $logger )
}
```

We'll need to make sure all logging lines use `WP_CLI::log()` instead of `WP_CLI::line()`

> Any ideas how to get around that? Is there a way we can find the "upgrade to" version and do something like And save STDOUT as {UPGRADE_VERSION}?

```
When I run `wp theme list --name=twentyfifteen --field=update_version`
Then save STDOUT as {UPDATE_VERSION}`
```

File: features/theme.feature [hunk 1]
```diff
@@ -249,3 +249,57 @@ Feature: Manage WordPress themes
...
+      """
+      Available theme updates:
+      Twenty Fifteen update from version 1.0 to version
+      """
+
+    When I run `wp theme update twentyfifteen --version=1.0`
+    Then STDOUT should not be empty
+
+    When I run `wp theme update twentyfifteen --format=json --dry-run`
+    Then STDOUT should contain:
```
Comment (danielbachhuber) [inline features/theme.feature]: Let's use `Then STDOUT should be JSON containing:`, which will ensure it's valid JSON.

File: features/theme.feature [hunk 2]
```diff
@@ -249,3 +249,57 @@ Feature: Manage WordPress themes
       | name          | status   |
       | biker         | active   |
       | jolene        | parent   |
+
+  Scenario: When updating a theme --format should be the same when using --dry-run
+    Given a WP install
+
+    When I run `wp theme update twentyfifteen --version=1.0`
```
Comment (danielbachhuber) [inline features/theme.feature]: Can we use `wp theme install --force twentyfifteen --version=1.0`, for versions of WordPress where `twentyfifteen` isn't installed by default? Also, we may want to use `twentytwelve` to ensure compatibility down to WP 3.7

File: features/theme.feature [hunk 3]
```diff
@@ -249,3 +249,57 @@ Feature: Manage WordPress themes
...
+    Then STDOUT should contain:
+      """
+      [{"name":"twentyfifteen","status":"inactive","version":"1.0",
+      """
+
+    When I run `wp theme update twentyfifteen --version=1.0`
+    Then STDOUT should not be empty
+
+    When I run `wp theme update twentyfifteen --format=csv --dry-run`
+    Then STDOUT should contain:
```
Comment (danielbachhuber) [inline features/theme.feature]: Same as above: `Then STDOUT should be CSV containing:`

File: features/theme.feature [hunk 4]
```diff
@@ -249,3 +249,57 @@ Feature: Manage WordPress themes
...
+      """
+
+  Scenario: Check json and csv formats when updating a theme
+    Given a WP install
+
+    When I run `wp theme update twentyfifteen --version=1.0`
+    Then STDOUT should not be empty
+
+    When I run `wp theme update twentyfifteen --format=json`
+    Then STDOUT should contain:
```
Comment (danielbachhuber) [inline features/theme.feature]: Same as above.

File: features/theme.feature [hunk 5]
```diff
@@ -249,3 +249,57 @@ Feature: Manage WordPress themes
...
+      """
+      [{"name":"twentyfifteen","old_version":"1.0",
+      """
+
+    When I run `wp theme update twentyfifteen --version=1.0`
+    Then STDOUT should not be empty
+
+    When I run `wp theme update twentyfifteen --format=csv`
+    Then STDOUT should contain:
```
Comment (danielbachhuber) [inline features/theme.feature]: Same as above.

File: php/WP_CLI/UpgraderSkin.php [hunk 6]
```diff
@@ -41,7 +41,9 @@ function feedback( $string ) {
 
 		$string = str_replace( '&#8230;', '...', strip_tags( $string ) );
 
-		\WP_CLI::line( $string );
+		if ( empty( $this->upgrader->cli_output_format ) || ! in_array( $this->upgrader->cli_output_format, array( 'csv', 'json' ) ) ) {
+			\WP_CLI::line( $string );
```
Comment (danielbachhuber) [inline php/WP_CLI/UpgraderSkin.php]: If we use `\WP_CLI::log()` here, we won't need to assign `$this->upgrader->cli_output_format`

File: php/WP_CLI/CommandWithUpgrade.php [hunk 7]
```diff
@@ -258,12 +266,14 @@ protected function update_many( $args, $assoc_args ) {
 
 		$line = "Updated $num_updated/$num_to_update {$this->item_type}s.";
 
-		if ( $num_to_update == $num_updated ) {
-			\WP_CLI::success( $line );
-		} else if ( $num_updated > 0 ) {
-			\WP_CLI::warning( $line );
-		} else {
-			\WP_CLI::error( $line );
+		if ( empty( $assoc_args['format'] ) || ! in_array( $assoc_args['format'], array( 'json', 'csv' ) ) ) {
```
Comment (danielbachhuber) [inline php/WP_CLI/CommandWithUpgrade.php]: Shouldn't need the `format` check here if we're using the Quiet logger.

**Comment by gilbitron** [comment]
@danielbachhuber Done. Only issue I had was using variables like `{UPDATE_VERSION}` in tables (`STDOUT should be CSV containing`) didn't seem to work.

**Comment by danielbachhuber** [comment]
:peach: Thanks!

**Pull request merged**